File: examples/mini-ts/src/ast.ts
// AST节点类型定义
export type BinaryOperator =
    'Add'
    | 'Subtract'
    | 'Multiply'
    | 'Divide'
    | 'Equal'
    | 'NotEqual'
    | 'LessThan'
    | 'LessThanOrEqual'
    | 'GreaterThan'
    | 'GreaterThanOrEqual';
export type UnaryOperator = 'Negate' | 'Not';

export type Type = 'number' | 'string' | 'boolean' | 'void';

export interface Literal {
    kind: 'Literal';
    type: Type;
    value: number | string | boolean;
}

export interface Identifier {
    kind: 'Identifier';
    name: string;
}

export interface BinaryExpression {
    kind: 'BinaryExpression';
    left: Expression;
    operator: BinaryOperator;
    right: Expression;
}

export interface UnaryExpression {
    kind: 'UnaryExpression';
    operator: UnaryOperator;
    operand: Expression;
}

export interface CallExpression {
    kind: 'CallExpression';
    callee: Identifier;
    arguments: Expression[];
}

export type Expression = Literal | Identifier | BinaryExpression | UnaryExpression | CallExpression;

export interface VariableDeclaration {
    kind: 'VariableDeclaration';
    name: string;
    type: Type;
    initializer?: Expression;
}

export interface FunctionDeclaration {
    kind: 'FunctionDeclaration';
    name: string;
    parameters: Parameter[];
    returnType: Type;
    body: Statement[];
}

export interface ReturnStatement {
    kind: 'ReturnStatement';
    value?: Expression;
}

export interface IfStatement {
    kind: 'IfStatement';
    condition: Expression;
    thenBranch: Statement[];
    elseBranch?: Statement[];
}

export interface WhileStatement {
    kind: 'WhileStatement';
    condition: Expression;
    body: Statement[];
}

export interface ExpressionStatement {
    kind: 'ExpressionStatement';
    expression: Expression;
}

export type Statement =
    | VariableDeclaration
    | FunctionDeclaration
    | ReturnStatement
    | IfStatement
    | WhileStatement
    | ExpressionStatement;

export interface Parameter {
    name: string;
    type: Type;
}

export interface Program {
    functions: FunctionDeclaration[];
}

// 将TypeScript类型转换为Gaia类型
export function typeToGaiaType(type: Type): string {
    switch (type) {
        case 'number':
            return 'F64'; // TypeScript的number对应Gaia的F64
        case 'string':
            return 'String';
        case 'boolean':
            return 'Bool';
        case 'void':
            return 'Void';
        default:
            throw new Error(`Unknown type: ${type}`);
    }
}

// 将字面量转换为Gaia常量
export function literalToGaiaConstant(literal: Literal): any {
    switch (literal.type) {
        case 'number':
            return {kind: 'Float', value: literal.value as number};
        case 'boolean':
            return {kind: 'Boolean', value: literal.value as boolean};
        case 'string':
            return {kind: 'String', value: literal.value as string};
        default:
            throw new Error(`Cannot convert literal of type ${literal.type} to Gaia constant`);
    }
}

File: examples/mini-rust/src/ast.rs
//! Mini Rust 抽象语法树定义

use serde::{Deserialize, Serialize};

/// Mini Rust 程序
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Program {
    pub name: String,
    pub functions: Vec<Function>,
}

/// 函数定义
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Function {
    pub name: String,
    pub parameters: Vec<Parameter>,
    pub return_type: Option<Type>,
    pub body: Block,
}

/// 函数参数
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Parameter {
    pub name: String,
    pub param_type: Type,
}

/// 代码块
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Block {
    pub statements: Vec<Statement>,
}

/// 语句
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Statement {
    /// 表达式语句
    Expression(Expression),
    /// 变量声明
    VariableDeclaration { name: String, var_type: Option<Type>, initializer: Option<Expression> },
    /// 返回语句
    Return(Option<Expression>),
}

/// 表达式
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Expression {
    /// 字面量
    Literal(Literal),
    /// 标识符
    Identifier(String),
    /// 函数调用
    FunctionCall { name: String, arguments: Vec<Expression> },
    /// 宏调用（如 println!）
    MacroCall { name: String, arguments: Vec<Expression> },
    /// 方法调用 (如 console.log)
    MethodCall { object: Box<Expression>, method: String, arguments: Vec<Expression> },
    /// 二元运算
    BinaryOperation { left: Box<Expression>, operator: BinaryOperator, right: Box<Expression> },
    /// 一元运算
    UnaryOperation { operator: UnaryOperator, operand: Box<Expression> },
}

/// 字面量
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Literal {
    /// 整数
    Integer(i64),
    /// 浮点数
    Float(f64),
    /// 字符串
    String(String),
    /// 布尔值
    Boolean(bool),
}

/// 二元运算符
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum BinaryOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
    Equal,
    NotEqual,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
}

/// 一元运算符
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum UnaryOperator {
    Negate,
    Not,
}

/// 类型
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Type {
    /// 32位整数
    I32,
    /// 64位整数
    I64,
    /// 32位浮点数
    F32,
    /// 64位浮点数
    F64,
    /// 字符串
    String,
    /// 布尔值
    Bool,
    /// 空类型
    Unit,
}

impl Type {
    /// 转换为 Gaia 类型
    pub fn to_gaia_type(&self) -> gaia_assembler::GaiaType {
        match self {
            Type::I32 => gaia_assembler::GaiaType::Integer32,
            Type::I64 => gaia_assembler::GaiaType::Integer64,
            Type::F32 => gaia_assembler::GaiaType::Float32,
            Type::F64 => gaia_assembler::GaiaType::Float64,
            Type::String => gaia_assembler::GaiaType::String,
            Type::Bool => gaia_assembler::GaiaType::Boolean,
            Type::Unit => gaia_assembler::GaiaType::Object, // 使用 Object 表示 unit 类型
        }
    }
}

impl Literal {
    /// 转换为 Gaia 常量
    pub fn to_gaia_constant(&self) -> gaia_assembler::GaiaConstant {
        match self {
            Literal::Integer(i) => gaia_assembler::GaiaConstant::Integer32(*i as i32),
            Literal::Float(f) => gaia_assembler::GaiaConstant::Float32(*f as f32),
            Literal::String(s) => gaia_assembler::GaiaConstant::String(s.clone()),
            Literal::Boolean(b) => gaia_assembler::GaiaConstant::Boolean(*b),
        }
    }
}
